File: 2024/day2/part2.lua
dofile("UsefulStuff.lua")


local function part2(fileName)

	local function report_valid(report)
		local inc, dec = true,true
		for i = 2,#report do
			local diff = report[i] - report[i-1]
			if math.abs(diff) > 3 or diff == 0 then
				inc,dec = false,false
				break
			elseif diff > 0 then
				dec = false
			else
				inc = false
			end
		end
		return inc or dec
	end

	-- Loop over each report
	local validReports = 0
	for line in io.lines(Script_path()..fileName) do

		-- Parse report into an array
		local report = {}
		for value in string.gmatch(line,"%d+") do
			table.insert(report, tonumber(value) )
		end

		-- Check validity, and if not try removing each item and seeing if the report
		-- becomes valid
		local valid = report_valid(report)
		for i = 1,#report do
			if valid then break end
			local reducedReport = Copy(report)
			table.remove(reducedReport,i)
			valid = report_valid(reducedReport)
		end
		if valid then
			validReports = validReports+1
		end
	end

	-- Return the number of valid reports
	return validReports
end

UnitTest(part2, "testcases/test1.txt",4)

local result = part2("input.txt")
assert(result == 364, "Wrong answer")
print("Part 2: " .. tostring(result))

File: 2025/day3/Javascript/Part2.js
const { assert } = require('console');
const fs = require('fs');

function part2(filepath) {
	banks = fs.readFileSync(filepath, 'UTF8').toString()
	.replaceAll(/\r\n/g,"\n").split('\n')
	.map(x =>  x.split('').map(y => Number(y)))
	
    total = 0
    banks.forEach(bank => {
		remaining = bank
		number = ''
		for (let i = 0; i < 12; i++) {
			digit = Math.max(...remaining.slice(0,(remaining.length-11+i)))
			number += digit
			remaining = remaining.slice(remaining.indexOf(digit)+1)
		}
		total+=Number(number)
    });
    return total
}

assert(part2(__dirname + '/../testcases/test1.txt') == 3121910778619)
assert(part2(__dirname + '/../input.txt') == 170520923035051)
console.log("Part 2:", part2(__dirname + '/../input.txt'))